#53-Implement ERD in Prisma Schema

eshwaribhide opened the pull request:
Closes #53

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+  DESIGN_ISSUE
+  NEW_FUNCTION
+  OTHER
+  STAGE_GATING
+  WP_INITIATION
```
I know you were saying something about having different tables for these (like Change Requests Table and then some other table), so I can add those or just keep it as is for now

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+}
+
+model Project {
+  project_id         Int            @id
```
Primary key of supertype becomes primary key of subtype, is what I believe when remembering what Wolfgang said. So this key should be the same as its companion in the WBS_Element table. However, when looking at the spotify project, it seems like, we made this key something else and then made the supertype key a foreign key. I can't seem to know which one it is and am not sure when googling. This also applies to the Work Package table.

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
-  age     Int?
-  posts   Post[]
-  profile Profile?
+  user_id           Int                @id @default(autoincrement())
```
I have added autoincrements to pretty much all IDs but I can remove them

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+
+model Work_Package {
+  wp_id        Int                  @id
+  wp_progress  Float
```
I made this a float because I don't think Prisma supports Reals, the closest thing then to needing a decimal value, for percentages such as wp progress, is Float

jamescd18 replied:
what do you think about float vs int? where the decimal place is doesn't matter to me

eshwaribhide replied:
We can make it a Float

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+  deliverables String
+  rules        String
+  desc_bullets Description_Bullet[]
+  project      Project              @relation(fields: [project_id], references: [project_id])
```
I was thinking about calling this parent_project and then something like parent_project_id or parent_proj_id but it just felt too long

jamescd18 replied:
meh. project and project id is good

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+  type            CR_Type
+  what            String
+  why             String
+  timeline_impact DateTime
+  budget_impact   Int
```
I changed the data types of these two, from text (String), as you mentioned

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+  project_id   Int
+}
+
+model Description_Bullet {
```
I think this also needs nullable fields.

jamescd18 reviewed (changes requested):
looks really good for a first iteration! I definitely think we wanna have better definition of our sub types and super types. Also, I think nullable fields is something missed in the schema. definitely make sure to update the ERD for any significant changes. this is super helpful in figuring out what the data model should look like. thank you so much!!!

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+  timeline_impact DateTime
+  budget_impact   Int
+  scope_impact    String
+  docs            String
```
Huh?

eshwaribhide commented:
Mr. Chang-Davidson, I think the ERD is fixed now, making Change Request a supertype and adding subtyping to reflect the different attributes that scope-based, stage gating, and wp initiation change requests will have...please let me know your comments as I might have missed something.

eshwaribhide commented:
I have added in the schema, the supertyping of change request and the corresponding subtypes (also unrelated but converted the default for fab weld request and change request date submitted to now() as you suggested).

I have some issues with the naming of things.

In regards to looking at comment syntax for Prisma, it's just "//////". Can you tell me where exactly you'd like me to add them as separators?

eshwaribhide commented:
Subtypes now have, as a foreign key, the key of their supertype. Also, I added a "review_notes" field in the Change Requests type and updated this also in the ERD.

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+  NEW_FUNCTION
+  OTHER
+  STAGE_GATING
+  WP_INITIATION
```
```suggestion
  WP_ACTIVATION
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+  DESIGN_ISSUE
+  NEW_FUNCTION
+  OTHER
+  STAGE_GATING
```
```suggestion
  STAGE_GATE
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+  ADMIN
+  PROJECT_MANAGER
+  PROJECT_LEAD
+  GUEST
```
```suggestion
  MEMBER
  GUEST
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
-  user   User    @relation(fields: [userId], references: [id])
-  userId Int     @unique
+enum Role {
+  ADMIN
```
```suggestion
  ADMIN
  LEADERSHIP
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+  wp_init        WP_Initiation[]
+}
+
+model Scope_Based_CR {
```
```suggestion
model Scope_CR {
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+  timeline_impact Int
+  budget_impact   Int
+  scope_impact    String
+  docs            String
```
```suggestion
  doc             String
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+
+model Description_Bullet {
+  desc_id      Int          @id @default(autoincrement())
+  detail       String?
```
```suggestion
  detail       String
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,170 @@ generator client {
+model Description_Bullet {
+  desc_id      Int          @id @default(autoincrement())
+  detail       String?
+  date_added   DateTime?
```
```suggestion
  date_added   DateTime
```

jamescd18 reviewed (changes requested):
Some good iteration and I clarified a few points.

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,122 @@ generator client {
+
+model Change {
+  change_id        Int            @id @default(autoincrement())
+  attribute        String
```
What would the enum types be? 'Create', 'update', and 'delete'?

eshwaribhide replied:
This has been addressed in commit 434e4113d7f43054a751ec41254200bca66b7d20. Also, commit 4466b33c9a3f796245be89e0be8f78c49b900959 adds the project_area field and a corresponding enum type for MechE or EE, and commit 1453389e103a84df7e5de76a933742af5c4b8803 adds the order in project field for work packages (an int made the most sense for the type, which we could then convert to a wbsNum).

jamescd18 reviewed (changes requested):
a few more changes

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,172 @@ generator client {
+  change_id        Int            @id @default(autoincrement())
+  attribute        String
+  original         String
+  new              Boolean
```
I have no idea why i wrote that

eshwaribhide commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,172 @@ generator client {
+  stage_gating_id    Int            @id
+  done               Boolean
+  leftover_budget    Int
+  design_reviewer    User           @relation(fields: [design_reviewer_id], references: [user_id])
```
I don't really think that is possible. The syntax above is for Prisma to know that this table will have a foreign key that is of type User, and in general, I don't think it's possible to have an array of foreign keys. I guess because this is a many to many relationship it might require having to make a Design Reviewer table mapping user ids to stage_gating_ids.

eshwaribhide replied:
Addressed by commit b386f59a908c023f8913ae7459f3e18155d71f58

eshwaribhide commented:
I have updated the ERD to reflect the latest changes

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,197 @@ generator client {
+model Work_Package {
+  wp_id            Int                  @id
+  order_in_project Int
+  wp_progress      Float
```
```suggestion
  wp_progress      Int
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,197 @@ generator client {
+  date_reviewed  DateTime
+  accepted       Boolean?
+  review_notes   String?
+  changes        Change[]
```
```suggestion
  changes        Change[]
  implemented        Boolean?
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,197 @@ generator client {
-  user   User    @relation(fields: [userId], references: [id])
-  userId Int     @unique
+enum Project_Area {
+  MECH_E
+  EE
```
```suggestion
  MECHANICAL
  ELECTRICAL
  BUSINESS
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,199 @@ generator client {
+  wbs_el_id      Int
+  Scope_CR       Scope_CR[]
+  stage_gate     Stage_Gate[]
+  wp_init        WP_Activation[]
```
```suggestion
  wp_activation        WP_Activation[]
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,199 @@ generator client {
+  role                       Role
+  Fab_Weld_Requests          Fab_Weld_Request[]
+  Change_Requests            Change_Request[]
+  WP_Init_Positions          WP_Activation[]
```
```suggestion
  WP_Activation_Positions          WP_Activation[]
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,199 @@ generator client {
+  GUEST
+}
+
+enum Change_Attr {
```
```suggestion
enum Change_Type {
```

jamescd18 commented on src/backend/prisma/schema.prisma:
```diff
@@ -10,29 +7,199 @@ generator client {
+
+model Change {
+  change_id        Int            @id @default(autoincrement())
+  attribute        Change_Attr
```
```suggestion
  attribute        Change_Type
```

jamescd18 reviewed (approved):
Good enough for now. Too complicated sometimes. Let's do it more iteratively.
